Add documentation on single file (#1709)
dotnet/diagnostics: documentation/single-file-tools.md

# Single File Diagnostic Tools

> All of our diagnostic tools have been distributed so far as [global tools](https://docs.microsoft.com/en-us/dotnet/core/tools/global-tools). These provide a great user experience as they provide clear ways to easily download, update, and configure the scope of the tools. However, these require a full .NET SDK which is often not available in some scenarios like production machines. For these scenarios we provide a single file distribution mechanism that only require a runtime to be available in the target machine. Other than the installation instructions, all functionality and usage remains the same. If you need help on how to use them, please refer to the [.NET Core diagnostic global tools](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/#net-core-diagnostic-global-tools) documentation.

Reviewer: ```suggestion
All of our diagnostic tools have been distributed so far as [global tools](https://docs.microsoft.com/en-us/dotnet/core/tools/global-tools). These provide a great user experience as they provide clear ways to easily download, update, and configure the scope of the tools. However, these require a full .NET SDK which is often not available in some scenarios like production machines. For these scenarios we provide a single file distribution mechanism that only requires a runtime to be available on the target machine. Other than the installation instructions, all functionality and usage remains the same. If you need help on how to use them, please refer to the [.NET Core diagnostic global tools](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/#net-core-diagnostic-global-tools) documentation.
```

## Requirements to Run the Tools

> - The tools distributed in a single file format are framework dependent applications. As such, they require a 3.1+ .NET Core runtime to be available. This means one of the following:

Reviewer: I would add something about customers using container images where the runtime is already installed globally. This is probably a common scenario where customers will want to acquire these tools.

  - The runtime was installed globally in a well-known location using one of the different installer technologies documented in the [official .NET download page](https://dotnet.microsoft.com/download).

>   - Downloading the binary archives from the [official .NET download page](https://dotnet.microsoft.com/download) and extracting them to a directory. Afterwards, please set `DOTNET_ROOT` to the path of the extraction.

Reviewer 2: ```suggestion
  - Downloading the binary archives from the [official .NET download page](https://dotnet.microsoft.com/download), extracting them to a directory, and setting `DOTNET_ROOT` to the extracted path.
```

Author: Ironically just did the same change 👍🏼

>   - Using the [dotnet-install scripts](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-install-script) and setting `DOTNET_ROOT` to the installation directory. This mechanism is not recommended for development or production environments, but it's ideal for environments like CI machines.

Reviewer 2: If we specifically are saying we don't recommend this mechanism, should we include _why_?

Author: It's in the dontet-install documentation. I don't believe it's pertinent here. I just wanted  to avoid people blindly using the mechanism that's not recommented

> - These tools are bundles that will inflate on the target machine and will extract files to disk as necessary. The extraction folder is by default within the `TEMP` directory. In case `TEMP` is not available in the target environment or you want to control such directory for reasons such as clean up, please set `DOTNET_BUNDLE_EXTRACT_BASE_DIR` to a path that can be used in lieu of `TEMP`.

Reviewer 2: ```suggestion
- These tools are self-extracting archives that will extract files to disk as necessary. The extraction target folder defaults to the platform's `TEMP` directory, e.g., the value of `$TMPDIR` on posix systems. In case `TEMP` is not available in the target environment or you want to control where the app is extracted to, you can set the environment variable `DOTNET_BUNDLE_EXTRACT_BASE_DIR` to a path that can be used in lieu of `TEMP`.
```

# Single File Diagnostic Tools

All of our diagnostic tools have been distributed so far as [global tools](https://docs.microsoft.com/en-us/dotnet/core/tools/global-tools). These provide a great user experience as they provide clear ways to easily download, update, and configure the scope of the tools. However, these require a full .NET SDK which is often not available in some scenarios like production machines. For these scenarios we provide a single file distribution mechanism that only require a runtime to be available in the target machine. Other than the installation instructions, all functionality and usage remains the same. If you need help on how to use them, please refer to the [.NET Core diagnostic global tools](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/#net-core-diagnostic-global-tools) documentation.

## Requirements to Run the Tools

- The tools distributed in a single file format are framework dependent applications. As such, they require a 3.1+ .NET Core runtime to be available. This means one of the following:

>   - The runtime was installed globally in a well-known location using one of the different installer technologies documented in the [official .NET download page](https://dotnet.microsoft.com/download).

Author: ```suggestion
  - Installing the runtime globally in a well-known location using one of the different installer technologies documented in the [official .NET download page](https://dotnet.microsoft.com/download).
```

>   - Downloading the binary archives from the [official .NET download page](https://dotnet.microsoft.com/download) and extracting them to a directory. Afterwards, please set `DOTNET_ROOT` to the path of the extraction.

Author: ```suggestion
  - Downloading the binary archives from the [official .NET download page](https://dotnet.microsoft.com/download), extracting them to a directory, and setting the `DOTNET_ROOT` environment variable to the path of the extraction.
```

  - Using the [dotnet-install scripts](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-install-script) and setting `DOTNET_ROOT` to the installation directory. This mechanism is not recommended for development or production environments, but it's ideal for environments like CI machines.

>   - The workload is being run in a container with the runtime available. This could be either because you are running a workload on our [official .NET runtime containers](https://hub.docker.com/_/microsoft-dotnet-runtime) or you built a custom image that installs it at build time.

Author: ```suggestion
  - The workload is being run in a container with the runtime available. This could be either because the workload is running on one of the [official .NET runtime containers](https://hub.docker.com/_/microsoft-dotnet-runtime) or on a custom-built image that installs it at build time.
```

> - These tools are bundles that will inflate on the target machine and will extract files to disk as necessary. The extraction folder is by default within the `TEMP` directory. In case `TEMP` is not available in the target environment or you want to control such directory for reasons such as clean up, please set `DOTNET_BUNDLE_EXTRACT_BASE_DIR` to a path that can be used in lieu of `TEMP`.

Author: ```suggestion
- These tools are bundles that will inflate on the target machine and will extract files to disk as necessary. The extraction folder is by default within the `TEMP` directory. In case `TEMP` is not available in the target environment or there's a desire to control such directory for reasons such as clean up, set `DOTNET_BUNDLE_EXTRACT_BASE_DIR` to a path that can be used in lieu of `TEMP`.
```

## Obtaining the Tools

### Downloading the latest available release of a tool

> The latest release of our tools is always available at a link that follows the following schema:

Author: ```suggestion
The latest release of there tools is always available at a link that follows the following schema:
```

```
https://aka.ms/<TOOL NAME>/<TARGET PLATFORM RUNTIME IDENTIFIER>
```

The tools we support are:
- `dotnet-counters`
- `dotnet-dump`
- `dotnet-gcdump`
- `dotnet-sos`
- `dotnet-trace` 
  
The supported [runtime identifiers](https://docs.microsoft.com/en-us/dotnet/core/rid-catalog) are:

- `win-arm`
- `win-arm64`
- `win-x64`
- `win-x86`
- `osx-x64`
- `linux-arm`.
- `linux-arm64`
- `linux-musl-arm64`
- `linux-x64`
- `linux-musl-x64`

> For example, the latest release of `dotnet-trace` for Windows x64 will be available at https://aka.ms/dotnet-dump/win-x64.

Author: ```suggestion
For example, the latest release of `dotnet-trace` for Windows x64 will be available at https://aka.ms/dotnet-trace/win-x64.
```

> To download these tools you can either use a browser or use a command line utility like `wget` or `curl`. It's important to note that the file name is returned in the `content-disposition` header of the request. If the utility you use to download the tools respects the header the file will be saved with the name of the tool downloaded, otherwise you might need to rename the tool. A few examples of ways to preserve the name:

Author: ```suggestion
To download these tools you can either use a browser or use a command line utility like `wget` or `curl`. It's important to note that the file name is returned in the `content-disposition` header of the request. If the utility you use to download the tools respects the header the file will be saved with the name of the tool downloaded, otherwise you might need to rename the tool. A few examples are:
```

1. *`curl` (available on Windows after 1706, macOS, and several distributions of Linux)*
```sh

> curl -JLO https://aka.ms/dotnet-dump/win-x64

Author: In the documentation folder? I might link them to this doc. Just adding a single link doesn't give much information on the requirements. I was trying to think of a way to combine them, but didn't see an easy way to caveat the requirements.

> curl -JLO https://aka.ms/dotnet-dump/win-x64

Reviewer 3: I'm ok whether you include it there or just link it to this doc - just want this to be more discoverable since those docs are where people usually land.

```

2. *`wget`*
```sh
wget --content-disposition https://aka.ms/dotnet-dump/linux-x64
```

3. *`pwsh` (PowerShell core)*
```powershell
$resp = Invoke-WebRequest -Uri "https://aka.ms/dotnet-dump/win-x86"
$header = [System.Net.Http.Headers.ContentDispositionHeaderValue]::Parse($resp.Headers.'content-disposition')
[System.IO.File]::WriteAllBytes($header.FileName, $resp.content)
```

> 

Author: ```suggestion

4. *`powershell`*
```powershell
Invoke-WebRequest -Uri "https://aka.ms/dotnet-dump/win-x86" -Outfile "dotnet-dump.exe"
```
```

### Past Releases and Checksum Validation

> Each release in the [releases section](https://github.com/dotnet/diagnostics/releases) of the repository contains a table of stable links for our tools starting release `v5.0.152202`.  Additionally, there's a CSV available as an attachment containing all the stable links and SHA512 checksums in case the downloaded file needs to be validated.

Author: ```suggestion
Each release in the [releases section](https://github.com/dotnet/diagnostics/releases) of the repository contains a table of stable links for our tools starting release `v5.0.152202`. Additionally, there's a CSV available as an attachment containing all the stable links and SHA512 checksums in case the downloaded file needs to be validated.
```

```

The tools we support are:

- `dotnet-counters`
- `dotnet-dump`
- `dotnet-gcdump`
- `dotnet-sos`
- `dotnet-trace`
  
The supported [runtime identifiers](https://docs.microsoft.com/en-us/dotnet/core/rid-catalog) are:

- `win-arm`
- `win-arm64`
- `win-x64`
- `win-x86`
- `osx-x64`
- `linux-arm`.
- `linux-arm64`
- `linux-musl-arm64`
- `linux-x64`
- `linux-musl-x64`

For example, the latest release of `dotnet-trace` for Windows x64 will be available at <https://aka.ms/dotnet-trace/win-x64>.

To download these tools you can either use a browser or use a command line utility like `wget` or `curl`. It's important to note that the file name is returned in the `content-disposition` header of the request. If the utility you use to download the tools respects the header, the file will be saved with the name of the tool downloaded; otherwise you might need to rename the tool. A few examples are:

1. *`curl` (available on Windows after 1706, macOS, and several distributions of Linux)*

```sh
curl -JLO https://aka.ms/dotnet-dump/win-x64
```

2. *`wget`*

```sh
wget --content-disposition https://aka.ms/dotnet-dump/linux-x64
```

3. *`pwsh` (PowerShell core)*

```powershell

> $resp = Invoke-WebRequest -Uri "https://aka.ms/dotnet-dump/win-x86"

Author: Powershell is only available on windows. I needed a windows link there. Wget is a downloadable tool, but it's not available by default on windows, so mostly that
